File: 2024/11/sol.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

var (
	countCache     = make(map[[2]int]int)
	transformCache = make(map[int][]int)
)

func count(stone int, blinks int) (res int) {
	key := [2]int{stone, blinks}
	if val, ok := countCache[key]; ok {
		return val
	}

	defer func() {
		countCache[key] = res
	}()

	if blinks == 0 {
		res = 1
		return
	}

	tvals := transform(stone)
	sum := 0
	for _, ns := range tvals {
		sum += count(ns, blinks-1)
	}
	res = sum
	return
}

func transform(stone int) (res []int) {
	if val, ok := transformCache[stone]; ok {
		return val
	}

	defer func() {
		transformCache[stone] = res
	}()

	if stone == 0 {
		res = []int{1}
		return
	}

	sstone := strconv.Itoa(stone)
	length := len(sstone)
	if length%2 == 0 {
		mid := length / 2
		leftPart := sstone[:mid]
		rightPart := sstone[mid:]

		leftVal, _ := strconv.Atoi(leftPart)
		rightVal, _ := strconv.Atoi(rightPart)
		res = []int{leftVal, rightVal}
		return
	}

	res = []int{stone * 2024}
	return
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	line := strings.TrimSpace(scanner.Text())
	parts := strings.Fields(line)

	var stones []int
	for _, p := range parts {
		val, _ := strconv.Atoi(p)
		stones = append(stones, val)
	}

	solution1 := 0
	solution2 := 0
	for _, s := range stones {
		solution1 += count(s, 25)
		solution2 += count(s, 75)
	}

	fmt.Println(solution1)
	fmt.Println(solution2)
}

File: 2024/11/sol.py
import sys
import functools


@functools.cache
def count(stone: str, blinks: int) -> int:
    if blinks == 0:
        return 1

    return sum(count(ns, blinks - 1) for ns in transform(stone))


@functools.cache
def transform(stone: int) -> list[int]:
    if stone == 0:
        return [1]

    sstone = str(stone)
    length = len(sstone)
    if length % 2 == 0:
        mid = length // 2
        left_part = sstone[:mid]
        right_part = sstone[mid:]

        return [int(left_part), int(right_part)]

    return [stone * 2024]


if __name__ == "__main__":
    line = sys.stdin.readline().strip()
    stones = [int(x) for x in line.split()]

    solution1 = 0
    solution2 = 0
    for s in stones:
        solution1 += count(s, 25)
        solution2 += count(s, 75)
    print(solution1)
    print(solution2)
